weave-js/src/components/PagePanelComponents/Home/Browse3/typeViews/moviepy.video.VideoClip/VideoPlayer.tsx

```diff
@@ -0,0 +1,188 @@
+import React, {useContext, useEffect, useState} from 'react';
+import {AutoSizer} from 'react-virtualized';
+
+import * as Dialog from '../../../../../../components/Dialog/Dialog';
+import VideoViewer from '../../../../../../components/Panel2/VideoViewer';
+import {LoadingDots} from '../../../../../LoadingDots';
+import {NotApplicable} from '../../NotApplicable';
+import {CustomLink} from '../../pages/common/Links';
+import {useWFHooks} from '../../pages/wfReactInterface/context';
+import {CustomWeaveTypePayload} from '../customWeaveType.types';
+import {CustomWeaveTypeProjectContext} from '../CustomWeaveTypeDispatcher';
+
+type VideoFormat = 'gif' | 'mp4' | 'webm';
+type VideoFileKeys = `video.${VideoFormat}`;
+
+type VideoClipTypePayload = CustomWeaveTypePayload<
+  'moviepy.video.VideoClip.VideoClip',
+  {[K in VideoFileKeys]: string}
+>;
+
+type VideoPlayerProps = {
+  entity: string;
+  project: string;
+  mode?: string;
+  data: VideoClipTypePayload;
+};
+
+const VIDEO_TYPES: Record<VideoFileKeys, VideoFormat> = {
+  'video.gif': 'gif',
+  'video.mp4': 'mp4',
+  'video.webm': 'webm',
+};
+
+const MIME_TYPES: Record<VideoFormat, string> = {
+  gif: 'image/gif',
+  mp4: 'video/mp4',
+  webm: 'video/webm',
+};
+
+export const VideoPlayer: React.FC<VideoPlayerProps> = props => (
+  <AutoSizer style={{height: '100%', width: '100%'}}>
+    {({width, height}) =>
+      width === 0 || height === 0 ? null : (
+        <VideoPlayerWithSize
+          {...props}
+          containerWidth={width}
+          containerHeight={height}
+        />
+      )
+    }
+  </AutoSizer>
+);
+
+type VideoPlayerWithSizeProps = VideoPlayerProps & {
+  containerWidth: number;
+  containerHeight: number;
+};
+
+const VideoPlayerWithSize: React.FC<VideoPlayerWithSizeProps> = ({
+  entity,
+  project,
+  data,
+  containerWidth,
+  containerHeight,
+  mode: propMode,
+}) => {
+  const [showPopup, setShowPopup] = useState(false);
+  const {useFileContent} = useWFHooks();
+  const context = useContext(CustomWeaveTypeProjectContext);
+  const mode = propMode ?? context?.mode;
+
+  // Find the first available video format
+  const videoKey = Object.keys(data.files).find(key => key in VIDEO_TYPES) as
+    | VideoFileKeys
+    | undefined;
+
+  // Always call the hook with consistent arguments
+  const videoBinary = useFileContent(
+    entity,
+    project,
+    videoKey ? data.files[videoKey] : '',
+    {skip: !videoKey}
+  );
+
+  if (!videoKey) {
+    return <NotApplicable />;
+  }
+
+  const fileExt = VIDEO_TYPES[videoKey];
+  const title = data.custom_name || entity.split('-')[0];
+
+  // Link mode (default view)
+  if (!mode || mode !== 'object_viewer') {
```

Comment:
I wonder if it would be wiser to have two components, one for link view and one for player view, instead of one with a `if`.
It seems the link needs much less code, so most of this component would be unused, which could become heavy in a long list of calls.

Reply:
That would be ideal. Unfortunately it's a bit difficult. The code is set up s.t. there's a single component which is used to render the CustomWeaveType. The only way we really have of controlling how this renders is by checking the `mode`.

This context already existed even though most components don't use it, so someone has definitely run into this issue before...